coordinator/src/node.rs

```diff
@@ -17,63 +17,108 @@ use std::sync::Arc;
-            trade_params.execution_price,
-            trade_params.quantity,
-            trade_params.leverage,
-        );
-
-        let total_collateral = margin_coordinator + margin_trader;
-
-        let maturity_time =
-            SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs() + trade_params.expiry.as_secs();
-
-        let contract_symbol = trade_params.contract_symbol.label();
-
-        // The contract input to be used for setting up the trade between the trader and the
-        // coordinator
-        let contract_input = ContractInput {
-            offer_collateral: margin_trader,
-            accept_collateral: margin_coordinator,
-            fee_rate: 2,
-            contract_infos: vec![ContractInputInfo {
-                contract_descriptor: dummy_contract_descriptor(total_collateral),
-                oracles: OracleInput {
-                    public_keys: vec![trade_params.oracle_pk],
-                    event_id: format!("{contract_symbol}{maturity_time}"),
-                    threshold: 1,
-                },
-            }],
-        };
+    /// Sets up the trade with the matched traders (maker, taker)
+    pub async fn trade(&self, match_params: MatchParams) -> Result<()> {
+        // todo: these proposals should be done in parallel with a timeout.
+        // propose trade to the taker
+        let taker_handle = self
+            .propose_trade(&match_params.taker, &match_params.params)
+            .await;
+
+        // propose trade to the maker
+        let maker_handle = self
+            .propose_trade(&match_params.maker, &match_params.params)
+            .await;
+
+        // todo: add proper error handling. In case on of the proposals fails, but the other
+        // succeeds we need to cancel the successful one.
+
+        taker_handle
+            .await
+            .context("Failed to wait on future")?
+            .context("Failed to propose trade to taker")?;
+
+        maker_handle
+            .await
+            .context("Failed to wait on future")?
+            .context("Failed to propose trade to maker")?;
+
+        Ok(())
+    }
+
+    pub async fn propose_trade(
+        &self,
+        trade: &Trade,
+        match_params: &Match,
+    ) -> JoinHandle<Result<()>> {
+        tokio::spawn({
+            let match_params = match_params.clone();
+            let trade = trade.clone();
+            let node = self.inner.clone();
+            async move {
+                let maturity_time = SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs()
+                    + match_params.expiry.as_secs();
+
+                let contract_symbol = match_params.contract_symbol.label();
+
+                let margin_trader = calculate_margin(
+                    match_params.execution_price,
+                    match_params.quantity as f64, // todo: the quantity should not be f64
+                    trade.leverage as f64,        // todo: the leverage should not be f64
```

Comment:
Unless we want to rule out fractional leverages this can be something else, otherwise I think this has to be a float.

Reply:
We *need* fractional leverage for a combined positions! This has to stay f64 from my point of view.

Reply:
thanks for the explanation.

Reply:
Can we use `Decimal` instead?

Reply:
> Can we use `Decimal` instead?

That might be a good idea!

Note: So far when using `Decimal`, we always use `float` as representation when serializing/deserializing it. It will still be better to use `Decimal` when performing calculations; but just for the sake of communicating information `f64` might actually be fine. We can however still use the `Decimal` type here, because then it becomes an implementation detail how it is serialized / desierialized.

Reply:
In the orderbook I've used `Decimal` in most places.